src/main/java/sparksoniq/spark/iterator/flowr/OrderByClauseSparkIterator.java

```diff
@@ -187,6 +195,105 @@ private void setAllLocalResults() {
 
     @Override
     public Dataset<Row> getDataFrame(DynamicContext context) {
-        return null;
+        if (this._child == null) {
+            throw new SparksoniqRuntimeException("Invalid orderby clause.");
+        }
+        Dataset<Row> df = _child.getDataFrame(context);
+        StructType inputSchema = df.schema();
+
+        df.sparkSession().udf().register("determineOrderingDataType",
+                new OrderClauseDetermineTypeUDF(_expressions, inputSchema),
+                DataTypes.createArrayType(DataTypes.StringType));
+
+        String udfSQL = DataFrameUtils.getUdfSQL(inputSchema);
+
+        df.createOrReplaceTempView("input");
+        Dataset<Row> columnTypesDf = df.sparkSession().sql(
+                String.format("select distinct(determineOrderingDataType(array(%s))) as `distinct-types` from input",
+                        udfSQL)
+        );
+
+        // Every column represents an order by expression
+        // Check that every column contains a single atomic type (nulls and empty-sequences are allowed)
+        Map<Integer, String> allColumnTypes = new LinkedHashMap<>();
+        Object columnTypesObject = columnTypesDf.collect();
+        Row[] columnTypesOfRows = ((Row[]) columnTypesObject);
+        for (Row columnTypesOfRow : columnTypesOfRows) {
+            List columnsTypesList = columnTypesOfRow.getList(0);
+            for (int columnIndex = 0; columnIndex < columnsTypesList.size(); columnIndex++) {
+                String columnType = (String) columnsTypesList.get(columnIndex);
+                if (!columnType.equals("empty-sequence") && !columnType.equals("null")) {
+                    if (allColumnTypes.get(columnIndex) == null) {
+                        allColumnTypes.put(columnIndex, columnType);
+                    } else if (!allColumnTypes.get(columnIndex).equals(columnType)) {
```

Comment:
Ah that's correct. I had started out with a 'numeric' implementation but while tackling various issues, I ended up defining 3 different types for numerics and forgetting the importance of conversion between them.

I have one question regarding handling of decimals and doubles though. Looking at the code fragment I have taken from **Item.java** class, the **getNumericResultType()** behaves differently compared to what you have described. The comment given is in line with your explanation, but it seems that implementations opts for decimals rather than doubles when the two are met together. Is this a bug ?

Lastly, as explained here https://stackoverflow.com/questions/3413448/double-vs-bigdecimal, BigDecimal aims to be more precise at the expense of performance. So in order to get a confirmation: Are we opting for performance rather than precision in this case, with the implementation you have described earlier, as we drop BigDecimals in favor of doubles when the two are met.
```
//performs conversions for binary operations with a numeric return type
//(int,double) -> double
//(int,decimal) -> decimal
//(decimal,double) -> double
public static Type getNumericResultType(Item left, Item right) {
    if (left instanceof DecimalItem)
        return DecimalItem.class;
    if (left instanceof DoubleItem) {
        if (right instanceof DecimalItem)
            return DecimalItem.class;
        return DoubleItem.class;
    }
    if (left instanceof IntegerItem) {
        if (right instanceof DoubleItem)
            return DoubleItem.class;
        if (right instanceof DecimalItem)
            return DecimalItem.class;
        return IntegerItem.class;
    }
    return DecimalItem.class;
}
```

Reply:
Very good question and indeed, this was a mistake in the code.

The W3C standard requires converting a decimal to a double if computed/compared with another double. This is called type promotion and it is specified here: https://www.w3.org/TR/xquery-30/#dt-type-promotion

Reply:
Perfect explanation! Glad to know that this was the intended behavior by the W3C standard. 

Following disclaimer in the docs wipes all my worries :)
> This kind of promotion may cause loss of precision.

I will fix the issue with implementation in this PR.